stan/math/rev/fun/fill.hpp

```diff
@@ -0,0 +1,65 @@
+#ifndef STAN_MATH_REV_FUN_FILL_HPP
+#define STAN_MATH_REV_FUN_FILL_HPP
+
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/rev/meta.hpp>
+#include <stan/math/rev/core.hpp>
+#include <vector>
+
+namespace stan {
+namespace math {
+
+/**
+ * Fill the specified container with the specified value.
+ *
+ * The specified matrix is filled by element.
+ *
+ * @tparam VarMat a `var_value` with inner type from `EigenBase`
+ * @tparam S A var.
+ *
+ * @param x Container.
+ * @param y Value.
+ */
+template <typename VarMat, typename S, require_var_matrix_t<VarMat>* = nullptr,
+          require_var_t<S>* = nullptr>
+inline void fill(VarMat& x, const S& y) {
+  arena_t<plain_type_t<value_type_t<VarMat>>> prev_vals = x.val().eval();
+  x.vi_->val_.fill(y.val());
+  if (!is_nan(y)) {
+    reverse_pass_callback([x, y, prev_vals]() mutable {
+      x.vi_->val_ = prev_vals;
+      y.adj() += x.adj().sum();
+      x.adj().setZero();
+    });
+  } else {
+    reverse_pass_callback(
+        [x, y, prev_vals]() mutable { x.vi_->val_ = prev_vals; });
```

Comment:
Add `x.adj().setZero();` if you decide it's important to keep the `is_nan` branch here. The previous entries of `x` should be unaffected by the NaN-ness of whatever replaces it.